### PR TITLE
Fix for __declspec(selectany) in clang 5.0

### LINKED ISSUE #4327: Chakracore not building with clang 5.0 on Linux
Building 1.7 or 1.8 from the source results in a large number of duplicate symbol errors, appearing to arise because instances of a const int are created in a header file that is included in multiple source files.   This is apparently tried to be mitigated by using a __declspec(selectany) construct to only link one of them, but this construct does not seem to work in Clang 5 on Linux.   While this may be a bug in the compiler, at the end of the day, it seems that there could still be workarounds to this which are less problematic.

Attached to this is a diff, which I would propose as a solution to the issue.   There are two components to the fix.  The first is to fix the ShiftValue in BVUnitT so that it did not complain about multiple definitions, and the second part is to change how MachPtr is defined in the architecture-dependent md.h file.  In particular, the const int construct is removed, and both replaced with enums, which can be evaluated at compile time.  Then, a template specialized constexpr function is used to specialize the value of ShiftValue, a regular global integer value is created in one source file to hold a duplicate of MachPtr that can be accessed externally, and the preprocessor is used to redirect access to what would otherwise be an extern const int to the global variable, so the change is transparent to everything else.

[chakracore.txt](https://github.com/Microsoft/ChakraCore/files/1506479/chakracore.txt)

### PR DESCRIPTION
**Pull request by gh-markt**
~~re: [issue 4327](https://github.com/Microsoft/ChakraCore/issues/4327)~~

Fixes #4327

### CODE REVIEW

File: lib/Backend/IRType.cpp [hunk 1]
```diff
@@ -5,6 +5,9 @@
 
 #include "Backend.h"
 
+#undef MachPtr
+int externMachPtr = MachPtr;
```
Comment (obastemur) [inline lib/Backend/IRType.cpp]: this feels quite error prone. (both previous state and now)

Can we move all `MachXXX` defs under IRType.h and `ifdef` based on archs?

**Comment by obastemur** [comment]
Notes on updating the PR;

- Make your changes on your local (on the branch `gh-markt:release/1.8`)
- see the list of files affected by your changes
```
git status
```
- add them
```
git add <path to the file>
```
- update your commit (the line below will open an editor.. save and exit)
```
git commit --amend
```
- push!
```
git push -f
```

**Comment by obastemur** [comment]
BTW; there are other `__declspec(selectany)` uses and they also fail. You might replace them with `extern`.

ie.
```
 #if defined(DBG) && !defined(DIAG_DAC)
 
 // AutoDebug functions that are only available in DEBUG builds
-_declspec(selectany) int AssertCount = 0;
-_declspec(selectany) int AssertsToConsole = false;
+extern int AssertCount;
+extern int AssertsToConsole;
 
 #if _WIN32
 _declspec(thread, selectany) int IsInAssert = false;
 #elif !defined(__IOS__)
-__declspec(thread, selectany) int IsInAssert = false;
+extern __declspec(thread) int IsInAssert;
 #else
 // todo: implement thread local variable for iOS ??
-__declspec(selectany) int IsInAssert = false;
+extern int IsInAssert;
 #endif
```

and define them under --- a/lib/Common/Core/Api.cpp
```
+int AssertCount = 0;
+int AssertsToConsole = false;
+#if !defined(__IOS__)
+__declspec(thread)
+#endif
+int IsInAssert = false;
```

**Comment by obastemur** [comment]
p.s. The reason `__declspec(selectany)` fails is a known Clang 5.0 bug.

File: lib/Backend/IRType.cpp (hunk 1 above)
Comment (gh-markt) [reply lib/Backend/IRType.cpp]: I will happily do so, buy as a more extensive change, I want to be up-front about concern for breakage to the architectures I cannot test on.

Comment (obastemur) [reply lib/Backend/IRType.cpp]: No worries, CI will test them all

**Comment by gh-markt** [comment]
I can alter these and update the pull request, but I would like to know if there is some mitigating reason why the above variables were not simply declared as thread_local?  thread_local is standard C++11, so which compiler was being supported that does not undertand thread_local?  I'm asking up front so that I know the right way to do this, and don't commit code that is not reflective of the the design intent.

**Comment by obastemur** [comment]
I would leave that definition as is. We got it already covered for cross platform and AFAIK this particular PR doesn’t need that change

**Comment by gh-markt** [comment]
Please advise if the request is acceptable, or if further changes are required.   Thank you.

**Comment by gh-markt** [comment]
Okay, now it's passing.  I'm not sure how to squash these commits into one after I've already done a pull request, however.   Please advise.

File: lib/Common/DataStructures/UnitBitVector.h [hunk 2]
```diff
@@ -69,8 +69,19 @@ GetLastBitSet(DWORD *Index, UnitWord64 Mask)
...
+
+template<> constexpr LONG BVUnitT_GetShiftValue<UnitWord32>() { return 5; }
+
+template<> constexpr LONG BVUnitT_GetShiftValue<UnitWord64>() { return 6; }
+
+}
+
 template <typename T>
-class BVUnitT
+class BVUnitT
```
Comment (obastemur) [inline lib/Common/DataStructures/UnitBitVector.h]: extra space

File: lib/Common/Core/Assertions.h [hunk 3]
```diff
@@ -9,17 +9,20 @@
 
 #if defined(DBG) && !defined(DIAG_DAC)
 
+
```
Comment (obastemur) [inline lib/Common/Core/Assertions.h]: nit: extra line

File: lib/Common/Core/Assertions.h [hunk 4]
```diff
@@ -9,17 +9,20 @@
...
 // AutoDebug functions that are only available in DEBUG builds
-_declspec(selectany) int AssertCount = 0;
-_declspec(selectany) int AssertsToConsole = false;
+extern int AssertCount;
+extern int AssertsToConsole;
+
 
+
```
Comment (obastemur) [inline lib/Common/Core/Assertions.h]: nit: extra line

File: lib/Common/Core/Api.cpp [hunk 5]
```diff
@@ -7,6 +7,19 @@
...
+#if _WIN32
+ // IsInAssert defined in header
+#elif !defined(__IOS__)
+ __declspec(thread) int IsInAssert = false;
+#else
+int IsInAssert = false;
+#endif
+
+
+
```
Comment (obastemur) [inline lib/Common/Core/Api.cpp]: nit: extra two lines

File: lib/Common/Core/Api.cpp [hunk 6]
```diff
@@ -7,6 +7,19 @@
 #include <wchar.h> // wmemcpy_s
 #endif
 
+int AssertCount = 0;
+int AssertsToConsole = false;
+
+#if _WIN32
+ // IsInAssert defined in header
```
Comment (obastemur) [inline lib/Common/Core/Api.cpp]: good one! Thanks

File: lib/Backend/IRType.h [hunk 7]
```diff
@@ -39,3 +37,14 @@ extern IRType IRType_EnsureUnsigned(IRType type);
 #if DBG_DUMP || defined(ENABLE_IR_VIEWER)
 extern void IRType_Dump(IRType type);
 #endif
+
+#ifdef _M_AMD64
+    #include "../../Backend/amd64/machvalues.h"
```
Comment (obastemur) [inline lib/Backend/IRType.h]: Did you try non-relative paths here? Both on *nix and Windows we define a good range of search path.

**Comment by obastemur** [comment]
Let's first give the final shape and eventually easy to squash them all. Thanks for the hard work here btw!

**Comment by gh-markt** [comment]
Please advise.   I don't appear to able to complete the signing off, although I'm not sure if that is my responsibility to do.   Clicking "details", and signing in, I receive the error message  "Sorry, but we’re having trouble with signing you in.", with an explanation "We've received a bad request".

If there are any other issues, please let me know.  Thanks.

File: lib/Backend/IRType.h (hunk 7 above)
Comment (gh-markt) [reply lib/Backend/IRType.h]: It seems as if the modules outside of lib/Backend don't have the architecture specific directory in their include path, so the specific path for the architecture has to be named explicitly.

I was therefore imitating the style used in lib/Backend/BackendOpCodeList.h, which does a similar arrangement of #includes for the appropriate platform for including MdOpCodes.h

Comment (obastemur) [reply lib/Backend/IRType.h]: `amd64/machvalues.h` instead of `../../Backend/amd64/machvalues.h` ? (same applies to others)

**Comment by obastemur** [comment]
@gh-markt I justed with Clang 5.0 and looks good.

Please update https://github.com/Microsoft/ChakraCore/pull/4335/files#r154331347 

Once this change above is done, at first, commit that as usual.

```
git commit -a -m "update this and that??"
```

Not it's time to squash them all;

```
git reset --soft HEAD~10
git commit --amend
git push -f
```

**Comment by obastemur** [comment]
fixed the comment. auto `correction` issues.

**Comment by gh-markt** [comment]
Okay, so how do I complete the sign off?  Or is there nothing left for me to do?  I have a microsoft online account already, but clicking on the details link and selecting my account to sign in, I get an error message "Sorry, but we’re having trouble with signing you in." with the explanation "We've received a bad request."

**Comment by obastemur** [comment]
LGTM. Thanks for the great work @gh-markt 

No worries, I will sign and merge.

**Comment by obastemur** [comment]
@dotnet-bot test all please

**Comment by obastemur** [comment]
There are couple of tiny white space issues, If CI passes I'm going to cherry-pick / cleanup white-space and merge.

@gh-markt Thanks for the contribution again.  If you have time, please take a look at the list of issues we need help -> https://github.com/Microsoft/ChakraCore/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22

**Comment by obastemur** [comment]
Merged via #4370 

@gh-markt welcome on board -> https://github.com/Microsoft/ChakraCore/graphs/contributors

**Pull request closed without merging**